DatadogCore/Sources/Core/Upload/BackgroundTaskCoordinator.swift

```diff
@@ -0,0 +1,30 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-Present Datadog, Inc.
+ */
+
+import Foundation
+
+/// The `BackgroundTaskCoordinator` protocol provides an abstraction for managing background tasks and includes methods for registering and ending background tasks.
+/// It serves as a useful abstraction for testing purposes as well as allows decoupling from UIKit in order to maintain Catalyst compliation.
+internal protocol BackgroundTaskCoordinator {
+    func registerBackgroundTask() -> BackgroundTaskIdentifier?
+    func endBackgroundTaskIfActive(_ backgroundTaskIdentifier: BackgroundTaskIdentifier)
+}
+
+internal typealias BackgroundTaskIdentifier = Int
+
+#if canImport(UIKit)
+import UIKit
+
+internal class UIKitBackgroundTaskCoordinator: BackgroundTaskCoordinator {
+    internal func registerBackgroundTask() -> BackgroundTaskIdentifier? {
+        return UIApplication.dd.managedShared?.beginBackgroundTask().rawValue
```

Comment:
https://developer.apple.com/documentation/uikit/uiapplication/1623031-beginbackgroundtask

> [Parameters](https://developer.apple.com/documentation/uikit/uiapplication/1623031-beginbackgroundtask#parameters)
handler
A handler to be called shortly before the app’s remaining background time reaches 0. Use this handler to clean up and mark the end of the background task. Failure to end the task explicitly will result in the termination of the app. The system calls the handler synchronously on the main thread, blocking the app’s suspension momentarily.

should there be a handler for cleanup?

Reply:
blocking/ 👍  At minimum we must call the `endBackgroundTask()` in cleanup handler to avoid crashing the app:

> Each call to this method must be balanced by a matching call to the [endBackgroundTask(_:)](https://developer.apple.com/documentation/uikit/uiapplication/1622970-endbackgroundtask) method. (...) If you don't call `endBackgroundTask(_:)` for each task before time expires, the system kills the app.

See Apple's example code [here](https://developer.apple.com/documentation/uikit/app_and_environment/scenes/preparing_your_ui_to_run_in_the_background/extending_your_app_s_background_execution_time#overview). Keep in mind that cleanup handler is told to be called on the main thread - this Apple's example code doesn't cover `self.backgroundTaskID` thread-safety.